### ISSUE TITLE
"Too many requests" error with ibm_cf causes hang with no explanation

### ISSUE THREAD

**Issue by LachlanStuart** [lithops-cloud/lithops#324]
I found that my code was frequently hanging in exactly the same spot, showing only 88 out of 256 activations having succeeded. The debug logs showed that only 88 activations successfully started, and there was no indication of what happened to the other 168 activations. 

I added a bunch of extra print statements throughout the code and found that the 168 activations took [this codepath](https://github.com/pywren/pywren-ibm-cloud/blob/1623bf5072564f402a04e59ace2b32f0efe3a669/pywren_ibm_cloud/libs/openwhisk/client.py#L228-L229).

The issue is that there's no external indication what went wrong. Hangs are not fun to debug - this one cost me over 4 hours. There are two places in this function that `return None`, which leads to a hang. I suggest that both places raise exceptions instead.

Debugging was also made difficult by [this exception handler](https://github.com/pywren/pywren-ibm-cloud/blob/master/pywren_ibm_cloud/executor.py#L355-L366) which doesn't reraise `KeyboardInterrupt`s, causing code to continue running after the exception, which makes it hard to get a good stack trace of where the code is hanging. There are several `except KeyboardInterrupt`s in the code that don't reraise - I suggest making sure that they all reraise the exception so that future debugging is easier.

**Comment by JosepSampe** [comment]
@LachlanStuart Those activations that took [this codepath](https://github.com/pywren/pywren-ibm-cloud/blob/1623bf5072564f402a04e59ace2b32f0efe3a669/pywren_ibm_cloud/libs/openwhisk/client.py#L228-L229) should not hang the execution. 429 Error indicates "To many requests in flight" This is not an error, this is a Quota limit, so in this particular, these invocations [go to a pending queue](https://github.com/pywren/pywren-ibm-cloud/blob/1623bf5072564f402a04e59ace2b32f0efe3a669/pywren_ibm_cloud/invoker.py#L216-L217) in order to be invoked at a later stage instead of stopping the whole execution. That's why it returns `None` instead of raising an exception.

Maybe the error is that when a failed invocation go to a pending queue, the async invoker is unable to retry the invocation.

**Comment by LachlanStuart** [comment]
@JosepSampe Ah, that makes sense. I didn't notice they were being put into a queue. It would be useful if this was logged.

I tried adding extra logging inside `_run_invoker_process` and [this was the result](https://gist.github.com/LachlanStuart/335935fa82fa03d4767b47671533c482) to figure out why it was hanging. It looks like it's getting stuck waiting for more tokens to be added to `token_bucket_q` when there have been multiple repeated 429 responses. Even after 20 minutes it hasn't resumed.

**Comment by JosepSampe** [comment]
Tokens are added to the `token_bucket_q` once an active function finishes its execution. Maybe the thread that is monitoring finished activations is not adding tokens to the queue. Can you add logging [here](https://github.com/pywren/pywren-ibm-cloud/blob/master/pywren_ibm_cloud/invoker.py#L385-L395)?

btw, what kind of job are you running? theoretically the quota limit is more than 1000 activations, and you get `429` only with 256 activations?

**Comment by LachlanStuart** [comment]
@JosepSampe Here are the logs with the extra logging: https://gist.github.com/LachlanStuart/ec3c7e218a5007771b90c3bc1a18cdb9

On today's run it happened first time I ran the pipeline, so it shouldn't have started with any existing API usage from the previous run. These were the jobs & timestamps:
```
2020-06-29 12:15:57,397 [INFO] ExecutorID 8bed6a/0 | JobID M000 - Starting function invocation: calculate_peaks_chunk()  - Total: 256 activations / 2048MB
2020-06-29 12:16:23,536 [INFO] ExecutorID 8bed6a/0 | JobID M001 - Starting function invocation: get_segm_stats()  - Total: 256 activations / 1024MB
2020-06-29 12:17:06,334 [INFO] ExecutorID 8bed6a/0 | JobID M002 - Starting function invocation: get_segm_stats()  - Total: 1 activations / 2048MB
2020-06-29 12:17:14,632 [INFO] ExecutorID 8bed6a/0 | JobID M003 - Starting function invocation: clip_centr_df_chunk()  - Total: 256 activations / 512MB
2020-06-29 12:17:31,130 [INFO] ExecutorID 8bed6a/0 | JobID M004 - Starting function invocation: get_first_peak_mz()  - Total: 256 activations / 512MB
2020-06-29 12:17:44,378 [INFO] ExecutorID 8bed6a/0 | JobID M005 - Starting function invocation: segment_centr_chunk()  - Total: 256 activations / 512MB
```

Only the last call received 429 warnings.

**Comment by JosepSampe** [comment]
I think I already understand where is the error. Thanks for your logs @LachlanStuart  :)

Just to put me in context, which region are you using? Do you always had these 429 errors? or they appeared recently?

**Comment by LachlanStuart** [comment]
@JosepSampe I'm using eu-de. I suspect it's something to do with my account setup, as I recently had to change accounts and didn't have these issues with previous accounts.

The default limit is supposed to allow 5000 API calls per minute, but in my case it seems like it might be set to only allow 600... Maybe they've changed the default without updating the documentation. Once you've fixed it I'll contact support, because apparently they can increase that limit. I just don't want to take away this opportunity to reproduce the issue, because I know we'll hit 5000 calls per minute at some point, and it'll be much harder to debug once we're running at that scale.

**Comment by JosepSampe** [comment]
Added fix in #330
